File: pb2.hs
-- 2 : Even Fibonacci numbers

-- sum_even_fib 4000000
-- 4613732
-- (107.04 secs, 31,117,087,152 bytes)

fib :: Int -> Int
fib 0 = 0
fib 1 = 1
fib n = fib (n -1) + fib (n -2)

sum_even_fib :: Int -> Int
sum_even_fib n = sum $ takeWhile (< n) [fib x | x <- [1..], even (fib x)]

File: pb2.py
"""Even Fibonacci numbers

By considering the terms in the Fibonacci sequence whose values do not exceed
four million, find the sum of the even-valued terms.
"""

def fibs():
    f1 = 0
    f2 = 1
    while True:
        f = f1 + f2
        yield f
        f1 = f2
        f2 = f


def sum_even_fib(n):
    s = 0
    for f in fibs():
        if f % 2 == 0:
            s += f
        if f >= n:
            break
    return s
